### PR TITLE
teach Sequelize's integer types to accept stringified integers

### PR DESCRIPTION
**Pull request by benesch**
@cdo it came to my attention today that our current Sequelize example's handling of int64s is horribly broken, since JS can't exactly represent integers returned by `unique_rowid()`. I have a patch to that repository coming soon, but this is a necessary prerequisite.

Going to upstream this as well, but figured I'd submit it here for now.

---

The JavaScript number type cannot represent all 64-bit integers--it can only exactly represent integers in the range [-2<sup>53</sup> + 1, 2<sup>53</sup> - 1]. Notably, CockroachDB's `unique_rowid()` function returns values outside the representable range.

This commit teaches Sequelize's `INTEGER` and `BIGINT` types to accept stringified numbers instead of just raw JavaScript numbers; it's otherwise impossible to store a number outside the representable range into a CockroachDB `INT` column.

### CODE REVIEW

**Comment by benesch** [comment]
Actually, now that I think about this, we may want to fix this in Cockroach itself. Looks like Postgres will happily convert a string literal to an int.

**Comment by benesch** [comment]
cc @bdarnell @jordanlewis

We've had [at least one user run into this problem](https://stackoverflow.com/questions/44692779/foreign-key-violations-with-serial-in-node-js/44739198#44739198) so far, and presumably there are more who didn't report it. Thoughts on landing this and cutting a release of `sequelize-cockroachdb`? That would make `int64`s immediately usable as foreign keys in Sequelize, even with CockroachDB v1.0.

**Comment by benesch** [comment]
@justinj, @itsbilal—@jordanlewis has suggested you two as additional reviewers.

File: index.js [hunk 1]
```diff
@@ -59,7 +61,30 @@ QueryGenerator.upsertQuery = function(tableName, insertValues, updateValues, whe
 
   console.log(insert + " ON CONFlICT (" + pkCols.join(',') + ") DO UPDATE SET " + onConflictSet + ";");
   return insert + " ON CONFlICT (" + pkCols.join(',') + ") DO UPDATE SET " + onConflictSet + ";";
```
Comment (justinj) [inline index.js]: lol `CONFlICT`, maybe you could fix that while you're here?

File: tests/helper.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+var chai = require('chai');
+var Sequelize = require('..');
+
+chai.use(require('chai-as-promised'));
+chai.use(require('chai-datetime'));
+
+before(function() {
+  this.sequelize = new Sequelize('sequelize_test', 'root', '', {
+    dialect: "postgres",
```
Comment (justinj) [inline tests/helper.js]: nit: you used single quotes elsewhere, but not here (I don't really care though, feel free to leave this)

File: index.js [hunk 3]
```diff
@@ -59,7 +61,30 @@ QueryGenerator.upsertQuery = function(tableName, insertValues, updateValues, whe
...
+// We must teach Sequelize's INTEGER and BIGINT types to accept stringified
+// numbers instead of just raw JavaScript numbers; it's otherwise impossible to
+// store a number outside the representable range into a CockroachDB INT column.
+[DataTypes.postgres.INTEGER, DataTypes.postgres.BIGINT].forEach(function (intType) {
+  // Disable escaping so that the returned string is not wrapped in quotes
+  // downstream. Valid integers cannot be dangerous, and we take care to reject
+  // invalid integers.
+  intType.prototype.escape = false;
+
+  intType.prototype.$stringify = function stringify(value) {
```
Comment (justinj) [inline index.js]: Where does `$stringify` come from? I can't find a reference to it in the Sequelize repo (only `stringify`)? Is it some wacky reflection/inheritance thing?

Comment (benesch) [inline index.js]: https://github.com/sequelize/sequelize/blob/v3/lib/data-types.js#L75-L80

(Sequelize's master branch renamed `$stringify` to something else.)

Comment (justinj) [inline index.js]: Ahh I see, I guess since that will be a major release we don't need to worry about this breaking?

Comment (benesch) [inline index.js]: Ugh, no, looks like v4 has come out since I opened this PR. Sigh.

**Pull request merged**